buildSrc/src/main/kotlin/CodegenTestCommon.kt

```diff
@@ -102,3 +107,135 @@ fun cargoCommands(properties: PropertyRetriever): List<Cargo> {
     }
     return ret
 }
+
+fun registerGenerateSmithyBuildTask(
+    rootProject: Project,
+    project: Project,
+    pluginName: String,
+    allCodegenTests: List<CodegenTest>
+) {
+    val properties = PropertyRetriever(rootProject, project)
+    project.tasks.register("generateSmithyBuild") {
+        description = "generate smithy-build.json"
+        outputs.file(project.projectDir.resolve("smithy-build.json"))
+
+        doFirst {
+            project.projectDir.resolve("smithy-build.json")
+                .writeText(
+                    generateSmithyBuild(
+                        rootProject.projectDir.absolutePath,
+                        pluginName,
+                        codegenTests(properties, allCodegenTests)
+                    )
+                )
+
+            // If this is a rebuild, cache all the hashes of the generated Rust files. These are later used by the
+            // `modifyMtime` task.
+            project.extra["previousBuildHashes"] = project.buildDir.walk()
+                .filter { it.name.endsWith(".rs") || it.name == "Cargo.toml" || it.name == "Cargo.lock" }
```

Comment:
This will work for our codegen, but technically Cargo treats files other than this as part of the source also since you can use the `include` macros to pull them into the binary. Basically any file that isn't git ignored is considered by cargo. I'm not really sure what cargo does in this case where the entire cargo project is git ignored since it's in the build directory.

Reply:
Yeah the filter is strictly unnecessary. Every file we generate for the crate is a candidate for having its `mtime` rolled back if we regenerate it with unchanged contents. Removed in https://github.com/awslabs/smithy-rs/pull/1422/commits/11dc09348b14dcb00837318646bd86b1f4e7fd2e.